pkg/util/ciutil/vars.go

```diff
@@ -34,36 +34,90 @@ type Vars struct {
+	PRNumber string
 }
 
 // DetectVars detects and returns the CI variables for the current environment.
 func DetectVars() Vars {
-	v := Vars{Name: DetectSystem()}
-	// All CI systems have a name (that's how we know we're in a CI system). After detecting one, we will
-	// try to detect some additional CI-specific metadata that the CLI will use. It's okay if we can't,
-	// we'll just have reduced functionality for our various CI integrations.
+	if os.Getenv("PULUMI_DISABLE_CI_DETECTION") != "" {
+		return Vars{Name: ""}
+	}
+
+	systemName := DetectSystem()
+	// If CI variables have been set specifically for Pulumi in the environment,
+	// use that in preference to attempting to automatically detect the CI system.
+	// This allows Pulumi to work with any CI system with appropriate configuration,
+	// rather than requiring explicit support for each one.
+	if os.Getenv("PULUMI_CI_SYSTEM") != "" {
+		v := Vars{}
+		// Override the name with the actual name in the env var.
+		v.Name = System(os.Getenv("PULUMI_CI_SYSTEM"))
+		v.BuildID = os.Getenv("PULUMI_CI_BUILD_ID")
+		v.BuildType = os.Getenv("PULUMI_CI_BUILD_TYPE")
+		v.BuildURL = os.Getenv("PULUMI_CI_BUILD_URL")
+		v.SHA = os.Getenv("PULUMI_CI_PULL_REQUEST_SHA")
+
+		// Don't proceed with automatic CI detection since we are using the PULUMI_* values.
+		return v
+	}
+
+	return detectCISystemVars(systemName)
+}
+
+// detectCISystemVars detects the various aspects of a CI environment
+// using pre-defined variables provided by each CI system.
+// It's okay if we can't detect the additional metadata,
+// we'll just have reduced functionality for our various CI integrations.
+func detectCISystemVars(systemName System) Vars {
+	v := Vars{Name: systemName}
+
 	switch v.Name {
 	case GitLab:
 		// See https://docs.gitlab.com/ee/ci/variables/.
 		v.BuildID = os.Getenv("CI_JOB_ID")
 		v.BuildURL = os.Getenv("CI_JOB_URL")
+		v.BuildType = os.Getenv("CI_PIPELINE_SOURCE")
 		v.SHA = os.Getenv("CI_COMMIT_SHA")
 		v.BranchName = os.Getenv("CI_COMMIT_REF_NAME")
 		v.CommitMessage = os.Getenv("CI_COMMIT_MESSAGE")
+		v.PRNumber = os.Getenv("CI_MERGE_REQUEST_ID")
 	case Travis:
 		// See https://docs.travis-ci.com/user/environment-variables/.
-		// NOTE: Travis doesn't set a build URL in its environment, https://github.com/travis-ci/travis-ci/issues/8935.
 		v.BuildID = os.Getenv("TRAVIS_JOB_ID")
 		v.BuildType = os.Getenv("TRAVIS_EVENT_TYPE")
+		v.BuildURL = os.Getenv("TRAVIS_BUILD_WEB_URL")
 		v.SHA = os.Getenv("TRAVIS_PULL_REQUEST_SHA")
 		v.BranchName = os.Getenv("TRAVIS_BRANCH")
 		v.CommitMessage = os.Getenv("TRAVIS_COMMIT_MESSAGE")
+		v.PRNumber = os.Getenv("TRAVIS_PULL_REQUEST")
 	case CircleCI:
 		// See: https://circleci.com/docs/2.0/env-vars/
 		v.BuildID = os.Getenv("CIRCLE_BUILD_NUM")
 		v.BuildURL = os.Getenv("CIRCLE_BUILD_URL")
 		v.SHA = os.Getenv("CIRCLE_SHA1")
 		v.BranchName = os.Getenv("CIRCLE_BRANCH")
+	case AzurePipelines:
```

Comment:
It's unfortunate that the case statements aren't already alphabetized, but we should definitely do that. Since otherwise it will become more and more gnarly as we add new CI/CD systems.

Also, same comment as before when setting multiple fields, let's keep them alphabetical as possible. Just to make it easier to visually scan/parse.

Reply:
I'll order them. But my hope is for something a little more elegant instead of having to expand this switch-case expression more. Something that I am planning to do with Bitbucket Pipelines env vars support.